crates/red_knot_python_semantic/resources/mdtest/stubs/class.md

```diff
@@ -6,7 +6,12 @@ In type stubs, classes can reference themselves in their base class definitions.
 `typeshed`, we have `class str(Sequence[str]): ...`.
 
 ```py path=a.pyi
-class C(C): ...
+class Foo[T]: ...
 
-reveal_type(C)  # revealed: Literal[C]
+# TODO we resolve `Foo[Bar]` to `Unknown` here without emitting a diagnostic.
+# (Ideally we'd understand generics, but failing that, we shouldn't *silently* infer `Unknown`)
```

Comment:
I was intrigued/confused by this TODO, because I didn't think we did silently infer `Unknown` for subscripts. And we don't! What's happening here is a different bug (authored by me!), where we silently ignore all diagnostics arising from deferred type inference (which includes class bases in a stub file.) I tracked it down and here is the fix, as a diff on this PR:

```diff
diff --git a/crates/red_knot_python_semantic/resources/mdtest/stubs/class.md b/crates/red_knot_pyt>
index cc1297499..28a6b1f5e 100644
--- a/crates/red_knot_python_semantic/resources/mdtest/stubs/class.md
+++ b/crates/red_knot_python_semantic/resources/mdtest/stubs/class.md
@@ -8,9 +8,8 @@ In type stubs, classes can reference themselves in their base class definitions.
 ```py path=a.pyi
 class Foo[T]: ...

-# TODO we resolve `Foo[Bar]` to `Unknown` here without emitting a diagnostic.
-# (Ideally we'd understand generics, but failing that, we shouldn't *silently* infer `Unknown`)
-class Bar(Foo[Bar]): ...
+# TODO subscripting generic types
+class Bar(Foo[Bar]): ...  # error: [non-subscriptable]

 reveal_type(Bar)  # revealed: Literal[Bar]
 reveal_type(Bar.__mro__)  # revealed: tuple[Literal[Bar], Unknown, Literal[object]]
diff --git a/crates/red_knot_python_semantic/src/types/infer.rs b/crates/red_knot_python_semantic/>
index 0c55db556..8e30049ee 100644
--- a/crates/red_knot_python_semantic/src/types/infer.rs
+++ b/crates/red_knot_python_semantic/src/types/infer.rs
@@ -410,15 +410,18 @@ impl<'db> TypeInferenceBuilder<'db> {
         if self.types.has_deferred {
             let mut deferred_expression_types: FxHashMap<ScopedExpressionId, Type<'db>> =
                 FxHashMap::default();
+            let mut deferred_diagnostics = TypeCheckDiagnostics::new();
             // invariant: only annotations and base classes are deferred, and both of these only
             // occur within a declaration (annotated assignment, function or class definition)
             for definition in self.types.declarations.keys() {
                 if infer_definition_types(self.db, *definition).has_deferred {
                     let deferred = infer_deferred_types(self.db, *definition);
                     deferred_expression_types.extend(&deferred.expressions);
+                    deferred_diagnostics.extend(&deferred.diagnostics);
                 }
             }
             self.types.expressions.extend(deferred_expression_types);
+            self.diagnostics.extend(&deferred_diagnostics);
         }

         self.check_class_definitions();
```

But if you'd rather not include the fix in this PR, that's totally reasonable; I can do it as a follow-up PR with a dedicated test.

Reply:
Nit: do we need the intermediate deferred types and diagnostics struct? I'm probably blind and there's an obvious borrow problem

Reply:
Heh, I was just looking at this and wondering the same thing! AFAICS there is no need for it; no idea why I did it that way. Updated diff:

```diff
diff --git a/crates/red_knot_python_semantic/resources/mdtest/stubs/class.md b/crates/red_knot_pyt>
index cc1297499..28a6b1f5e 100644
--- a/crates/red_knot_python_semantic/resources/mdtest/stubs/class.md
+++ b/crates/red_knot_python_semantic/resources/mdtest/stubs/class.md
@@ -8,9 +8,8 @@ In type stubs, classes can reference themselves in their base class definitions.
 ```py path=a.pyi
 class Foo[T]: ...

-# TODO we resolve `Foo[Bar]` to `Unknown` here without emitting a diagnostic.
-# (Ideally we'd understand generics, but failing that, we shouldn't *silently* infer `Unknown`)
-class Bar(Foo[Bar]): ...
+# TODO subscripting generic types
+class Bar(Foo[Bar]): ...  # error: [non-subscriptable]

 reveal_type(Bar)  # revealed: Literal[Bar]
 reveal_type(Bar.__mro__)  # revealed: tuple[Literal[Bar], Unknown, Literal[object]]
diff --git a/crates/red_knot_python_semantic/src/types/infer.rs b/crates/red_knot_python_semantic/>
index 0c55db556..4ca286be0 100644
--- a/crates/red_knot_python_semantic/src/types/infer.rs
+++ b/crates/red_knot_python_semantic/src/types/infer.rs
@@ -408,17 +408,15 @@ impl<'db> TypeInferenceBuilder<'db> {
         }

         if self.types.has_deferred {
-            let mut deferred_expression_types: FxHashMap<ScopedExpressionId, Type<'db>> =
-                FxHashMap::default();
             // invariant: only annotations and base classes are deferred, and both of these only
             // occur within a declaration (annotated assignment, function or class definition)
             for definition in self.types.declarations.keys() {
                 if infer_definition_types(self.db, *definition).has_deferred {
                     let deferred = infer_deferred_types(self.db, *definition);
-                    deferred_expression_types.extend(&deferred.expressions);
+                    self.types.expressions.extend(&deferred.expressions);
+                    self.diagnostics.extend(&deferred.diagnostics);
                 }
             }
-            self.types.expressions.extend(deferred_expression_types);
         }

         self.check_class_definitions();
```